examples/server/server.cpp

```diff
@@ -1,46 +1,116 @@
-#include <httplib.h>
-#include <json.hpp>
 #include "common.h"
 #include "llama.h"
+#include "build-info.h"
+
+#include "httplib.h"
+#include "json.hpp"
 
 struct server_params
 {
   std::string hostname = "127.0.0.1";
   int32_t port = 8080;
+  int32_t read_timeout = 600;
+  int32_t write_timeout = 600;
+  bool verbose = false;
+};
+
+static size_t common_part(const std::vector<llama_token> & a, const std::vector<llama_token> & b) {
+  size_t i;
+  for (i = 0; i < a.size() && i < b.size() && a[i] == b[i]; i++);
+  return i;
+}
+
+enum stop_type {
+    STOP_FULL,
+    STOP_PARTIAL,
 };
 
+bool ends_with(const std::string &str, const std::string &suffix)
+{
+    return str.size() >= suffix.size() &&
+           0 == str.compare(str.size() - suffix.size(), suffix.size(), suffix);
+}
+
+size_t find_partial_stop_string(const std::string &stop, const std::string &text)
+{
+    if (!text.empty()) {
+        const char text_last_char = text.back();
+        for (int64_t char_index = stop.size() - 1; char_index >= 0; char_index--) {
+            if (stop[char_index] == text_last_char) {
+                const std::string current_partial = stop.substr(0, char_index + 1);
+                if (ends_with(text, current_partial)) {
+                    return text.size() - char_index - 1;
+                }
+            }
+        }
+    }
+    return std::string::npos;
+}
+
+static std::string debug_str(const std::string & s) {
```

Comment:
This function is named `debug_str` but to me it looks more like special characters are being escaped.

Reply:
It is used for the verbose logging. But I guess it could be renamed.